### ISSUE TITLE
Optimize memory usage in level_dataset.py

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#204]
In anticipation of having larger datasets, it will be more important to optimize the way they are stored in memory, and I have several suggestions from Claude, which you can see here:

"1. Preprocess into numpy once, not JSON every run. Convert scenes into a numpy integer array. If your levels are fixed-size (or you're already bucketing to a handful of shapes), store one array per bucket shape, or a single (N, H, W) int8 array with padding + a small mask/shape-index array. If truly ragged, store a flat 1D buffer of tile ids plus an offsets array (classic ragged-array trick) — still numpy, no per-tile Python objects. This alone fixes both the memory bloat and the DataLoader-worker duplication problem, and it also fixes something you haven't mentioned yet but will hit: JSON parsing time. At 1M samples your JSON file will be multiple GB of text, and json.load will take real, noticeable time (and a memory spike while parsing) every single time you start a training run or restart after a crash. A one-time preprocessing script that dumps to .npy/.npz (or np.memmap if you don't want it all in RAM) turns that into a near-instant load."

Basically, this is proposing an extra step in our pipeline. I want to retain backwards compatibility by allowing users to provide a json file for training,  but now they should also have the option to provide a .npy or .npz file instead. Also, even if a json is provided, the level_dataset should convert the data at load time into numpy array(s) before training. We can have an option that makes the level_dataset automatically create and also save to disk the numpy version of the json dataset so that it will be automatically used if training stops and resumes, and is generally available if we need it.

**Comment by fwilson12** [comment]
Everything should be in place; see the commit details above for more info, but essentially we can now process our .json ([ {scene: ..., caption: ..} ...]) datasets into a (N, H, W) .npy array and a metadata-only .json, which *significantly* reduces the net memory usage:

<img width="709" height="86" alt="Image" src="https://github.com/user-attachments/assets/555a5ab9-81cf-4a62-8120-890ae6638e33" />

##

I ran back to back tiny test runs in train_diffusion to make sure the initial caching and retrieval worked as it should, but haven't started a full training run yet. I'll close the issue for now, but come back to it if anything weird happens down the road

**Comment by schrum2** [comment]
Go ahead and do a generic Mario training test run, using the standard train-conditional.bat file to test things out. 

In fact, can you add the same timing logs you did for Mega Man to the Mario batch files? That way we can properly track the performance.

**Comment by fwilson12** [comment]
After running the batch file twice, I've confirmed that subsequent train_diffusion calls will load cached .npy datasets in place of their .json counterparts (even if the datasets argued with --json and --val_json point to the old .json datasets, the script will determine if the same filename has a .npy counterpart and load from there instead if it exists)


CLI output from the second run after the datasets were cached in the first run:
>Using loss function: COMBO
Loaded text encoder from Mar1and2-MLM-regular0
No block embedding model specified. One-hot encoding enabled.
Loading cached numpy bundle datasets\Mar1and2_LevelsAndCaptions-regular-train.scenes.npy (source: datasets\Mar1and2_LevelsAndCaptions-regular-train.json)...
Training samples: 6918
remove_upside_down_pipes: False
Loading cached numpy bundle datasets\Mar1and2_LevelsAndCaptions-regular-validate.scenes.npy (source: datasets\Mar1and2_LevelsAndCaptions-regular-validate.json)...
Training samples: 384

**Comment by fwilson12** [comment]
I'll let the run go for a little bit and make sure the samples look good before closing the issue

**Comment by fwilson12** [comment]
Here are samples after 60 epochs, looks like everything is working:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/e9802835-5ad2-4c7f-833f-f70fe18e1b90" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/549dd128-8259-46d6-8843-f463920084d3" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/7643c24d-0a9a-420c-af0b-0998ae1e9ab7" />

##

Caption adherence is also good. 
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/1d8f6bed-3576-4707-b800-1dd31c66e1ae" />
